## 빙어 낚시터를 방문시 잡을 수 있는 물고기의 예상 크기는?

In [2]:
import numpy as np

In [12]:
# 가상의 데이터를 만들기위하여 도메인 조사
fish = np.arange(6, 11)
print(type(fish))
print(fish.shape)
print(fish)

<class 'numpy.ndarray'>
(5,)
[ 6  7  8  9 10]


In [13]:
tot = sum(fish)
print(tot)

40


In [5]:
mean = tot / len(fish)
print(mean)

7.0


In [16]:
np.random.seed(1)
fish = np.random.randint(6, 11, size=100)  # 빙어 100 마리 
print(type(fish))
print(fish)

<class 'numpy.ndarray'>
[ 9 10  6  7  9  6  6  7 10 10  7  8 10  8 10  9 10  8 10  8 10  7  7  6
  7  7  7  7  6 10  7  6  6  9  8  7  6  9  7  7  9 10  6  7  9 10  8 10
  6  9  7  8  6 10  7  8  8  7  6  7  9 10  9  7  9  6  6  8  8  7  9 10
  8  6  6  7  7  9  6  6 10  8 10  9  9  6  9 10  9 10 10 10  7  6 10  8
  6  8 10  7]


In [18]:
tot = np.sum(fish)
print(tot)

797


In [19]:
mean = np.mean(fish)
print(mean)

7.97


In [20]:
print('예상되는 물고기의 크기:', mean, 'cm')
print('기대되는 물고기의 크기:', mean, 'cm') # 기대값
print('대표적으로 많이 잡히는 물고기의 크기:', mean, 'cm')

예상되는 물고기의 크기: 7.97 cm
기대되는 물고기의 크기: 7.97 cm
대표적으로 많이 잡히는 물고기의 크기: 7.97 cm


In [21]:
# 7.97 cm의 물고기만 잡히는 않음, 편차: 평균으로부터의 차이
for item in fish[0:10]:
    print(mean-item) # 개별 항목에 대한 편차, 편차 = 평균 - 데이터

-1.0300000000000002
-2.0300000000000002
1.9699999999999998
0.9699999999999998
-1.0300000000000002
1.9699999999999998
1.9699999999999998
0.9699999999999998
-2.0300000000000002
-2.0300000000000002


In [11]:
# 평균으로부터의 차이(편차)는 0이 나와야 정확히 빙어의 크기를 예측 가능하다.
# 편차가 작아야 기대값이 정확하다.

In [25]:
# 100마리를 잡은 사람이 평균 7.97 cm라고하는데 모든 차이를 알고싶다.
fish_devi = 0; # 편차의 합
for item in fish:  
    fish_devi = fish_devi + (mean-item) # 개별 항목에 대한 편차의 누적 합
    
print('편차의 합: {0:.1f}'.format(fish_devi))    
# 100마리의 빙어는 모두 7.97 cm로 편차가 없다.

편차의 합: -0.0


In [26]:
# 편차의 합은 왜 0?
src = np.arange(-2, 6)
src

array([-2, -1,  0,  1,  2,  3,  4,  5])

In [27]:
m = np.mean(src)
print('mean:', m)
for item in src:
    print(m-item)

mean: 1.5
3.5
2.5
1.5
0.5
-0.5
-1.5
-2.5
-3.5


In [28]:
hap = 0
for item in src:
    hap = hap + (m-item)
    
print('편차의 합:', hap)    

편차의 합: 0.0


In [16]:
# 100마리의 빙어는 모두 7.97 cm로 편차가 없다.
# 편차가 존재하는데 편차가 없다는 논리적 모순에 빠짐.

In [29]:
# 편차의 부호를 변경하여 모두 더함.
fish_devi = 0  # 편차
fish_devi_square = 0; # 편차의 합
for item in fish:
    fish_devi = mean - item
    fish_devi_square = fish_devi_square + (fish_devi ** 2)

print('평균:', mean)    
print('편차의 합: {0:.1f}'.format(fish_devi_square)) 

평균: 7.97
편차의 합: 216.9


In [30]:
# 평균 편차를 알고 싶음, 분산, 2.191 cm
var = fish_devi_square / (100-1)     # 표본분산 : -1 하는것
# 분산: 데이터들이 평균(대표값)으로부터 얼마나 떨어져 있는지의 척도
var

2.191010101010102

In [19]:
# 인하대학교 앞에 김밥천국 인수시 분산이 너무 큰 경우 발견
# 3월 ~ 6월, 9월 ~ 11월만 장가사 잘됨, 방학때 매출액이 매우 낮음
# 일정하지 않은 매출액으로 운영에 어려움이 많아 인수 포기.

In [31]:
# 분산은 음수를 양수로 변경하기위해 편차를 제곱했음으로 실제 편차가 아님.
print(np.sqrt(4))
print(np.sqrt(9))
print(np.sqrt(16))
print(np.sqrt(10))
sd = np.sqrt(var)
print(sd) # 빙어 잡을 시 상황에 따라 약 1.4802 ** 2를 2.191에 근사함(근접함)

2.0
3.0
4.0
3.1622776601683795
1.4802061008555876


In [21]:
print('평균:', mean)    # 실제 편차가 제곱되어 있음
print('분산:', var)    # 실제 편차가 제곱되어 있음
print('표준편차:', sd) # 제곱을 해제함.
print('예상 빙어의 크기:', mean-sd,' ~',mean+sd)

평균: 6.97
분산: 2.1691000000000007
표준편차: 1.472786474679884
예상 빙어의 크기: 5.497213525320116  ~ 8.442786474679885


In [22]:
fish

array([8, 9, 5, 6, 8, 5, 5, 6, 9, 9, 6, 7, 9, 7, 9, 8, 9, 7, 9, 7, 9, 6,
       6, 5, 6, 6, 6, 6, 5, 9, 6, 5, 5, 8, 7, 6, 5, 8, 6, 6, 8, 9, 5, 6,
       8, 9, 7, 9, 5, 8, 6, 7, 5, 9, 6, 7, 7, 6, 5, 6, 8, 9, 8, 6, 8, 5,
       5, 7, 7, 6, 8, 9, 7, 5, 5, 6, 6, 8, 5, 5, 9, 7, 9, 8, 8, 5, 8, 9,
       8, 9, 9, 9, 6, 5, 9, 7, 5, 7, 9, 6])

In [33]:
np.var(fish)

2.1691

In [32]:
print(np.std(fish)) # 모표준편차
print(np.std(fish, ddof=1)) # 표본표준편차


1.4727864746798838

## 통계에서의 분산과 표준편차
- 실제 데이터 분석시 모집단의 모든 데이터를 수집하지 못하는 경우가 일반적임.
- A 친구가 잡은 빙어는 그 호수에서 잡은 전체 빙어중 극히 일부임.
- 그렇다면 그 친구의 분산과 표준편차를 신뢰하기는 어렵다.
- 100마리의 빙어에 대해서만 산출된 데이터
- 평균: 6.97
- 분산: 2.1691000000000007
- 표준편차: 1.472786474679884
- 예상 빙어의 크기: 5.497213525320116  ~ 8.442786474679885

In [25]:
1/5

0.2

In [26]:
1/4  # 분모의 수가 줄어들면 나누기 값은 증가함

0.25

In [27]:
# 호수 전체(모집단)에서의 평균
print('mean:', mean)
print('편차의 합: {0:.3f}'.format(fish_devi_square)) 
sample_var = fish_devi_square / (100-1)  # n-1
print('모분산: {0:.3f}'.format(var)) # 전체 데이터 인식
print('표본 분산: {0:.3f}'.format(sample_var)) # 부분데이터로 인식

mean: 6.97
편차의 합: 216.910
모분산: 2.169
표본 분산: 2.191


In [30]:
# 모든 데이터를 가지고 있는 우리는 모분산을 알고 있으니 일반적인
# 분석 환경에서는 모분산을 알 수 없는 경우가 대부분임.
std = np.sqrt(var)
sample_std = np.sqrt(sample_var)
print('모표준편차: {0:.3f}'.format(std)) # 전체 데이터 인식
print('표본 표준편차: {0:.3f}'.format(sample_std)) # 부분데이터로 인식

모표준편차: 1.473
표본 쵸준편차: 1.480


In [32]:
fish = [i for i in range(-2, 3)]
print(fish)

[-2, -1, 0, 1, 2]


In [38]:
print('합계:', np.sum(fish))
print('모분산:', np.var(fish))
print('모분산:', np.var(fish, ddof=0))        # n-ddof -> 5 - 0 -> 5
print('표본 모분산:', np.var(fish, ddof=1))   # n-ddof -> 5 - 1 -> 4
print('모표준편차:', np.std(fish, ddof=0))    # n-ddof -> 5 - 0 -> 5
print('표본 표준편차:', np.std(fish, ddof=1)) # n-ddof -> 5 - 1 -> 4

합계: 0
모분산: 2.0
모분산: 2.0
표본 모분산: 2.5
모표준편차: 1.4142135623730951
표본 표준편차: 1.5811388300841898


### 표본 분산, 표본 표준편차의 실습

In [42]:
# 모분산: 모든 모집단의 수를 대상으로 함
# 표본분산: 모집단의 모든 수를 대상으로 분산을 구하기 어려움으로 표본을
# 이용하여 분산을 구하고 마치 모든 데이터를 대상으로 한것처럼 사용함.
# 표본은 모집단의 성질을 가지고 있다고 통계적으로 이론이 정립되어있음.
# 호수 전체에는 1 ~ 20cm의 20마리의 물고기가 있음.
parent = np.arange(1, 21)
print(parent)
parent_vari = np.var(parent)
print('호수 전체의 모분산:', parent_vari)  #  ddof=0
print('호수 전체를 표본으로본 표본 분산:', np.var(parent, ddof=1))

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]
호수 전체의 모분산: 33.25
호수 전체를 표본으로본 표본 분산: 35.0


In [47]:
# 첫번째 낚시꾼
fisher1 = np.arange(1, 5)
print(fisher1)
fisher1_vari = np.var(fisher1, ddof=1)
print('첫번째 낚시꾼이 잡은 물고기의 표본 분산:', fisher1_vari)

[1 2 3 4]
첫번째 낚시꾼이 잡은 물고기의 표본 분산: 1.6666666666666667


In [48]:
# 두번째 낚시꾼
fisher2 = np.arange(5, 11)
print(fisher2)
fisher2_vari = np.var(fisher2, ddof=1)
print('두번째 낚시꾼이 잡은 물고기의 표본 분산:', fisher2_vari)

[ 5  6  7  8  9 10]
두번째 낚시꾼이 잡은 물고기의 표본 분산: 3.5


In [49]:
# 세번째 낚시꾼
fisher3 = np.arange(11, 16)
print(fisher3)
fisher3_vari = np.var(fisher3, ddof=1)
print('세번째 낚시꾼이 잡은 물고기의 분산:', fisher3_vari)

[11 12 13 14 15]
세번째 낚시꾼이 잡은 물고기의 분산: 2.5


In [50]:
# 네번째 낚시꾼
fisher4 = np.arange(16, 21)
print(fisher4)
fisher4_vari = np.var(fisher4, ddof=1)
print('네번째 낚시꾼이 잡은 물고기의 분산:', fisher4_vari)

[16 17 18 19 20]
네번째 낚시꾼이 잡은 물고기의 분산: 2.5


In [ ]:
# 4명의 낚시꾼은 호수의 물고기 20마리를 전부 잡는데 성공함.

In [53]:
# 4명의 낚시꾼이 잡은 물고기의 분산
fisher_vari_tot = fisher1_vari+fisher2_vari+fisher3_vari+fisher4_vari
print('4명의 낚시꾼이 잡은 물고기의 표본 분산의 합:', fisher_vari_tot)
print('호수 전체의 분산:', parent_vari)

4명의 낚시꾼이 잡은 물고기의 표본 분산의 합: 10.166666666666668
호수 전체의 분산: 33.25


In [54]:
# 표본의 경우 데이터가 모집단보다 작음으로 물리적으로 거리도 적게 떨어짐
# 따라서 분산의 값도 모집단을 분할하여 소규모 데이터로 나눈후 분산값을 더해도
# 모집단보다 작아지는 통계량이 발생함.
# 해결방법은 데이터의 갯수를 n-1하여 1개를 줄임으로 분산의 값을 증가시킴

In [57]:
# 첫번째 낚시꾼 표본 분산
fisher1 = np.arange(1, 5)
print(fisher1)
fisher1_mean = np.mean(fisher1)
fisher1_divi = (fisher1 - fisher1_mean) ** 2 # 편차의 제곱
print(fisher1_divi)
fisher1_divi_tot = np.sum(fisher1_divi) # 편차의 제곱의 합
print('편차의 제곱의 합:', fisher1_divi_tot)
fisher1_vari = fisher1_divi_tot / (len(fisher1)-1)
print('첫번째 낚시꾼이 잡은 물고기의 표본 분산:', fisher1_vari)

[1 2 3 4]
[2.25 0.25 0.25 2.25]
편차의 제곱의 합: 5.0
첫번째 낚시꾼이 잡은 물고기의 표본 분산: 1.6666666666666667


In [59]:
# 두번째 낚시꾼 표본 분산
fisher2 = np.arange(5, 11)
print(fisher2)
fisher2_mean = np.mean(fisher2)
fisher2_divi = (fisher2 - fisher2_mean) ** 2 # 편차의 제곱
print(fisher2_divi)
fisher2_divi_tot = np.sum(fisher2_divi) # 편차의 제곱의 합
print('편차의 제곱의 합:', fisher2_divi_tot)
fisher2_vari = fisher2_divi_tot / (len(fisher2)-1)
print('두번째 낚시꾼이 잡은 물고기의 표본 분산:', fisher2_vari)

[ 5  6  7  8  9 10]
[6.25 2.25 0.25 0.25 2.25 6.25]
편차의 제곱의 합: 17.5
두번째 낚시꾼이 잡은 물고기의 표본 분산: 3.5
